### ISSUE TITLE
cargo check caching incorrect result

### ISSUE THREAD

**Issue by bjorn3** [rust-lang/rust-analyzer#16565]
This caused a fair amount of confusion in https://users.rust-lang.org/t/writing-a-client-for-rust-analyzer/106810

Basically the issue is that if you create a new cargo project and put `fn main() { 1; }` in src/main.rs, run the below node.js script which tries to get diagnostics from rust-analyzer, then put `fn main() { 1 + ""; }` in src/main.rs, and run the script again, the pass result is cached as such no diagnostics are reported. Even `cargo check` incorrectly passes.

```js
const { spawn } = require('child_process');
const path = require('path');

// Path to the Rust Analyzer executable and the file to analyze
const rustAnalyzerPath = '/home/bjorn/.var/app/com.visualstudio.code/data/vscode/extensions/rust-lang.rust-analyzer-0.4.1841-linux-x64/server/rust-analyzer';
const filePath = path.resolve('src/main.rs');

// Start the Rust Analyzer server
const server = spawn(rustAnalyzerPath, [], { stdio: ['pipe', 'pipe', process.stderr] });

// Function to send a JSON-RPC message to the server
function sendMessage(method, params, id) {
    const message = JSON.stringify({ jsonrpc: '2.0', method, params, id });
    const contentLength = Buffer.byteLength(message, 'utf8');
    server.stdin.write(`Content-Length: ${contentLength}\r\n\r\n${message}`);
}

// Function to send a JSON-RPC request to the server
function sendRequest(method, params, id) {
    sendMessage(method, params, id);
}

// Function to send a JSON-RPC notification to the server
function sendNotification(method, params) {
    sendMessage(method, params, null);
}


// Initialize the LSP client
sendRequest('initialize', {
    processId: process.pid,
    rootPath: path.dirname(path.dirname(filePath)),
    rootUri: `file://${path.dirname(path.dirname(filePath))}`,
    capabilities: {},
}, 1);

// Send the initialized notification
sendNotification('initialized', {});

// Open the file
sendNotification('textDocument/didOpen', {
    textDocument: {
        uri: `file://${filePath}`,
        languageId: 'rust',
        version: 1,
        text: require('fs').readFileSync(filePath, 'utf8'),
    },
});

// Read the server's responses
server.stdout.on('data', (data) => {
    console.log('Server response:', data.toString());
});
```

**rust-analyzer version**: (eg. output of "rust-analyzer: Show RA Version" command, accessible in VSCode via <kbd>Ctrl/⌘</kbd>+<kbd>Shift</kbd>+<kbd>P</kbd>)

**rustc version**: rustc 1.76.0 (07dca489a 2024-02-04)

**relevant settings**: The node.js script lists all used settings.

cc @Hofer-Julian

**Comment by Veykril** [comment]
That's probably the rustc_wrapper causing havoc again. Does that happen with `rust-analyzer.cargo.buildScripts.useRustcWrapper` disabled?

**Comment by bjorn3** [comment]
It doesn't happen with the rustc wrapper disabled.

@Hofer-Julian you can add `initializationOptions: { "cargo": { "buildScripts": { "useRustcWrapper": false } } }` to the initialize request to pass this option. Just be aware that for larger projects it can cause the initial load to take longer as the entire project will be built when this option is disabled and not just the build scripts and proc macros.

**Comment by Hofer-Julian** [comment]
I can confirm, that the caching issue doesn't occur with `{ "useRustcWrapper": false }`. I also experience that rust-analyzer needs significantly more time until it gives back results.
Thanks for the help @Veykril and @bjorn3!

**Comment by Veykril** [comment]
Closing as duplicate https://github.com/rust-lang/rust-analyzer/issues/14722
